// route-lite/build/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.goBack = exports.goTo = exports.Link = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var StackItem = function StackItem(component, props) {
  _classCallCheck(this, StackItem);

  this.component = component;
  this.props = props;
};

var stack = new Array();
var component = void 0;

function goTo(comp) {
  var props = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  stack.push(new StackItem(comp, props));
  component && component.forceUpdate();
}

function goBack() {
  if (stack.length > 0) {
    stack.pop();
  }
  component && component.forceUpdate();
}

var Link = exports.Link = function (_React$PureComponent) {
  _inherits(Link, _React$PureComponent);

  function Link(props) {
    _classCallCheck(this, Link);

    var _this = _possibleConstructorReturn(this, (Link.__proto__ || Object.getPrototypeOf(Link)).call(this, props));

    _this.onClick = function (evt) {
      evt.preventDefault();
      if (_this.props.component) {
        goTo(_this.props.component, _this.props.componentProps || {});
      } else if (_this.props.href) {
        open(_this.props.href);
      }
      if (typeof _this.props.onClick === 'function') {
        _this.props.onClick(evt);
      }
    };

    return _this;
  }

  _createClass(Link, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var dataProps = {};
      Object.keys(this.props).forEach(function (key) {
        if (key.match('data')) {
          dataProps[key] = _this2.props[key];
        }
      });

      return _react2.default.createElement(
        'a',
        _extends({
          href: this.props.href ? this.props.href : '',
          className: this.props.className,
          id: this.props.id
        }, dataProps, {
          onClick: this.onClick
        }),
        this.props.children
      );
    }
  }]);

  return Link;
}(_react2.default.PureComponent);

Link.defaultProps = {
  className: '',
  id: ''
};

var Router = function (_React$Component) {
  _inherits(Router, _React$Component);

  function Router() {
    _classCallCheck(this, Router);

    var _this3 = _possibleConstructorReturn(this, (Router.__proto__ || Object.getPrototypeOf(Router)).call(this));

    _this3.componentDidMount = _this3._componentDidMount.bind(_this3);
    return _this3;
  }

  _createClass(Router, [{
    key: '_componentDidMount',
    value: function _componentDidMount() {
      component = this;
    }
  }, {
    key: 'render',
    value: function render() {
      var _ref = stack[stack.length - 1] || {},
          _ref$component = _ref.component,
          Component = _ref$component === undefined ? function (_ref2) {
        var children = _ref2.children;
        return children;
      } : _ref$component,
          props = _ref.props;

      return _react2.default.createElement(
        Component,
        props,
        stack.length == 0 && this.props.children
      );
    }
  }]);

  return Router;
}(_react2.default.Component);

exports.default = Router;
exports.goTo = goTo;
exports.goBack = goBack;
