// fetchai-ledger-api/dist/fetchai/ledger/api/server.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ServerApi = void 0;

var _common = require("./common");

var _axios = _interopRequireDefault(require("axios"));

var _errors = require("../errors");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class ServerApi extends _common.ApiEndpoint {
  async status() {
    // Gets the status of a constellation server
    let url = `${this.protocol()}://${this.host()}:${this.port()}/api/status`;
    let response;

    try {
      response = await (0, _axios.default)({
        method: 'get',
        url: url
      });
    } catch (error) {
      throw new _errors.ApiError('Malformed response from server');
    }

    return response.data;
  }

  async num_lanes() {
    // Queries the ledger for the number of lanes currently active
    const status = await this.status();
    return status.lanes;
  }

  async version() {
    const status = await this.status();
    return status.version;
  }

}

exports.ServerApi = ServerApi;
