// fetchai-ledger-api/dist/contracts/transfer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NO_INIT = exports.MULTIPLE_INITS = exports.SIMPLE_CONTRACT = exports.TRANSFER_CONTRACT = void 0;
const TRANSFER_CONTRACT = `
persistent sharded balance_state : UInt64;

@init
function setup(owner : Address)
  use balance_state[owner];

  balance_state.set(owner, 1000000u64);
endfunction

@action
function transfer(from: Address, to: Address, amount: UInt64)
  use balance_state[from, to];

  // Check if the sender has enough balance to proceed
  if (balance_state.get(from, 0u64) >= amount)
    // update the account balances
    balance_state.set(from, balance_state.get(from) - amount);
    balance_state.set(to, balance_state.get(to, 0u64) + amount);
  endif

endfunction

@query
function balance(address: Address) : UInt64
  use balance_state[address];

  return balance_state.get(address, 0u64);
endfunction

`;
exports.TRANSFER_CONTRACT = TRANSFER_CONTRACT;
const SIMPLE_CONTRACT = `
@init
function init(owner: Address)
endfunction

@action
function action1()
endfunction

@action
function action2()
endfunction

@query
function query1()
endfunction

@query
function query2()
endfunction
`;
exports.SIMPLE_CONTRACT = SIMPLE_CONTRACT;
const MULTIPLE_INITS = `
@init
function setup(owner: Address)
endfunction

@init
function alternative(owner: Address)
endfunction
`;
exports.MULTIPLE_INITS = MULTIPLE_INITS;
const NO_INIT = `
@action
function action1()
endfunction
`;
exports.NO_INIT = NO_INIT;
